File: benchmarks/dormant_vs_polling.py
#!/usr/bin/env python3
"""
Benchmark: Spikeling's event-driven "dormant brain" model vs a naive
always-polling FSM, at varying NPC counts and varying %-active-near-player.

This tests the one specific architectural claim that underlies the
"practical/profitable game AI" pitch for Spikeling: that only simulating
NPC brains when they're near the player (event-driven / dormant) is
cheaper than the common naive approach of running every NPC's full
decision logic every tick regardless of relevance.

Both sides do REAL work per NPC, not stubs:
  - Polling FSM: every NPC, every tick, runs a small decision-tree style
    check (distance compare, health compare, line-of-sight stand-in,
    state transition) -- representative of typical naive game AI.
  - Spikeling brain: a 5-neuron LIF network (same shape as
    fps-game/enemy_brain.spk) is only INTEGRATED (the actual neuron math)
    for NPCs within activation range. NPCs outside range still need a
    cheap distance check to know they're dormant -- that part is NOT
    free, and is included in both models' cost for a fair comparison.

Run it:
    python dormant_vs_polling.py
"""

import math
import random
import time


# ---------- shared cheap distance check (paid by both models) ----------

def distance_check(npc_pos, player_pos):
    dx = npc_pos[0] - player_pos[0]
    dy = npc_pos[1] - player_pos[1]
    return math.sqrt(dx * dx + dy * dy)


# ---------- naive polling FSM: full logic every NPC, every tick ----------

class PollingNPC:
    __slots__ = ("pos", "health", "state")

    def __init__(self, pos):
        self.pos = pos
        self.health = 100.0
        self.state = "idle"

    def tick(self, player_pos, activation_range):
        dist = distance_check(self.pos, player_pos)
        # naive FSM: always runs the full check regardless of distance
        if dist < activation_range:
            if self.health < 30:
                self.state = "flee"
            elif dist < activation_range * 0.3:
                self.state = "attack"
            else:
                self.state = "chase"
        else:
            self.state = "idle"
        # small fixed per-tick cost representing "is doing something" logic
        self.health -= 0.0001
        return self.state


# ---------- Spikeling-style dormant LIF brain ----------

class LIFNeuron:
    __slots__ = ("threshold", "leak", "potential")

    def __init__(self, threshold, leak):
        self.threshold = threshold
        self.leak = leak
        self.potential = 0.0

    def step(self, drive):
        self.potential += drive
        self.potential -= self.leak
        if self.potential < 0:
            self.potential = 0.0
        fired = self.potential >= self.threshold
        if fired:
            self.potential = 0.0
        return fired


class SpikelingNPC:
    __slots__ = ("pos", "health", "brain", "state")

    def __init__(self, pos):
        self.pos = pos
        self.health = 100.0
        self.state = "dormant"
        # 5-neuron brain, same shape as fps-game/enemy_brain.spk
        # (SIGHT, DAMAGE, CHASE, RECOIL, ATTACK)
        self.brain = [
            LIFNeuron(threshold=1.0, leak=0.05),
            LIFNeuron(threshold=1.0, leak=0.05),
            LIFNeuron(threshold=1.0, leak=0.05),
            LIFNeuron(threshold=1.0, leak=0.05),
            LIFNeuron(threshold=1.0, leak=0.05),
        ]

    def tick(self, player_pos, activation_range):
        dist = distance_check(self.pos, player_pos)  # paid regardless, same as polling model
        if dist >= activation_range:
            self.state = "dormant"
            return self.state  # brain NOT simulated -- this is the whole optimization

        # only active NPCs actually integrate their neurons
        sight_drive = max(0.0, 1.0 - dist / activation_range)
        fired = [n.step(sight_drive) for n in self.brain]
        if fired[1]:
            self.state = "flee"
        elif fired[4]:
            self.state = "attack"
        elif fired[2]:
            self.state = "chase"
        else:
            self.state = "alert"
        self.health -= 0.0001
        return self.state


# ---------- benchmark harness ----------

def run_trial(npc_count, active_fraction, ticks, model):
    activation_range = 50.0
    player_pos = (0.0, 0.0)

    npcs = []
    n_active = int(npc_count * active_fraction)
    for i in range(npc_count):
        if i < n_active:
            r = random.uniform(0, activation_range * 0.8)
        else:
            r = random.uniform(activation_range * 1.5, activation_range * 5)
        angle = random.uniform(0, 2 * math.pi)
        pos = (r * math.cos(angle), r * math.sin(angle))
        npcs.append(model(pos))

    start = time.perf_counter()
    for _ in range(ticks):
        for npc in npcs:
            npc.tick(player_pos, activation_range)
    elapsed = time.perf_counter() - start
    return elapsed


def main():
    ticks = 200
    npc_counts = [50, 200, 1000]
    active_fractions = [0.05, 0.20, 0.50, 1.00]

    print(f"{'NPCs':>6} {'%active':>8} {'Polling FSM (s)':>17} {'Spikeling LIF (s)':>19} {'Speedup':>9}")
    print("-" * 64)
    for npc_count in npc_counts:
        for frac in active_fractions:
            t_poll = run_trial(npc_count, frac, ticks, PollingNPC)
            t_spike = run_trial(npc_count, frac, ticks, SpikelingNPC)
            speedup = t_poll / t_spike if t_spike > 0 else float("inf")
            print(f"{npc_count:>6} {frac*100:>7.0f}% {t_poll:>17.4f} {t_spike:>19.4f} {speedup:>8.2f}x")
        print()


if __name__ == "__main__":
    main()

File: benchmarks/dormant_vs_polling.c
/*
 * C version of dormant_vs_polling.py -- same NPC counts, same active
 * fractions, same 5-neuron LIF shape, same polling FSM logic. The only
 * variable changed vs the Python version is the language/runtime, to
 * isolate whether interpreter overhead was the reason the dormant model
 * lost to naive polling in Python.
 *
 * Build:
 *   gcc -O2 -o dormant_vs_polling dormant_vs_polling.c -lm
 * Run:
 *   ./dormant_vs_polling
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

typedef struct { double x, y; } Vec2;

static double distance_check(Vec2 a, Vec2 b) {
    double dx = a.x - b.x, dy = a.y - b.y;
    return sqrt(dx * dx + dy * dy);
}

/* ---------- polling FSM NPC ---------- */

typedef struct {
    Vec2 pos;
    double health;
    int state; /* 0=idle 1=flee 2=attack 3=chase */
} PollingNPC;

static void polling_tick(PollingNPC *n, Vec2 player_pos, double activation_range) {
    double dist = distance_check(n->pos, player_pos);
    if (dist < activation_range) {
        if (n->health < 30.0) n->state = 1;
        else if (dist < activation_range * 0.3) n->state = 2;
        else n->state = 3;
    } else {
        n->state = 0;
    }
    n->health -= 0.0001;
}

/* ---------- Spikeling dormant LIF NPC ---------- */

typedef struct {
    double threshold, leak, potential;
} LIFNeuron;

static int lif_step(LIFNeuron *neu, double drive) {
    neu->potential += drive;
    neu->potential -= neu->leak;
    if (neu->potential < 0.0) neu->potential = 0.0;
    int fired = neu->potential >= neu->threshold;
    if (fired) neu->potential = 0.0;
    return fired;
}

typedef struct {
    Vec2 pos;
    double health;
    int state; /* 0=dormant 1=flee 2=attack 3=chase 4=alert */
    LIFNeuron brain[5];
} SpikelingNPC;

static void spikeling_init(SpikelingNPC *n, Vec2 pos) {
    n->pos = pos;
    n->health = 100.0;
    n->state = 0;
    for (int i = 0; i < 5; i++) {
        n->brain[i].threshold = 1.0;
        n->brain[i].leak = 0.05;
        n->brain[i].potential = 0.0;
    }
}

static void spikeling_tick(SpikelingNPC *n, Vec2 player_pos, double activation_range) {
    double dist = distance_check(n->pos, player_pos);
    if (dist >= activation_range) {
        n->state = 0;
        return; /* brain NOT simulated -- the actual optimization */
    }
    double sight_drive = 1.0 - dist / activation_range;
    if (sight_drive < 0.0) sight_drive = 0.0;
    int fired[5];
    for (int i = 0; i < 5; i++) fired[i] = lif_step(&n->brain[i], sight_drive);
    if (fired[1]) n->state = 1;
    else if (fired[4]) n->state = 2;
    else if (fired[2]) n->state = 3;
    else n->state = 4;
    n->health -= 0.0001;
}

/* ---------- benchmark harness ---------- */

static double frand(double lo, double hi) {
    return lo + (hi - lo) * ((double)rand() / RAND_MAX);
}

static double run_polling(int npc_count, double active_fraction, int ticks) {
    double activation_range = 50.0;
    Vec2 player_pos = {0.0, 0.0};
    PollingNPC *npcs = malloc(sizeof(PollingNPC) * npc_count);
    int n_active = (int)(npc_count * active_fraction);
    for (int i = 0; i < npc_count; i++) {
        double r = (i < n_active) ? frand(0, activation_range * 0.8)
                                   : frand(activation_range * 1.5, activation_range * 5);
        double angle = frand(0, 2 * M_PI);
        npcs[i].pos.x = r * cos(angle);
        npcs[i].pos.y = r * sin(angle);
        npcs[i].health = 100.0;
        npcs[i].state = 0;
    }
    clock_t start = clock();
    for (int t = 0; t < ticks; t++)
        for (int i = 0; i < npc_count; i++)
            polling_tick(&npcs[i], player_pos, activation_range);
    double elapsed = (double)(clock() - start) / CLOCKS_PER_SEC;
    free(npcs);
    return elapsed;
}

static double run_spikeling(int npc_count, double active_fraction, int ticks) {
    double activation_range = 50.0;
    Vec2 player_pos = {0.0, 0.0};
    SpikelingNPC *npcs = malloc(sizeof(SpikelingNPC) * npc_count);
    int n_active = (int)(npc_count * active_fraction);
    for (int i = 0; i < npc_count; i++) {
        double r = (i < n_active) ? frand(0, activation_range * 0.8)
                                   : frand(activation_range * 1.5, activation_range * 5);
        double angle = frand(0, 2 * M_PI);
        Vec2 pos = { r * cos(angle), r * sin(angle) };
        spikeling_init(&npcs[i], pos);
    }
    clock_t start = clock();
    for (int t = 0; t < ticks; t++)
        for (int i = 0; i < npc_count; i++)
            spikeling_tick(&npcs[i], player_pos, activation_range);
    double elapsed = (double)(clock() - start) / CLOCKS_PER_SEC;
    free(npcs);
    return elapsed;
}

int main(void) {
    srand(42);
    int ticks = 20000;
    int npc_counts[] = {50, 200, 1000, 10000};
    double fractions[] = {0.05, 0.20, 0.50, 1.00};

    printf("%6s %8s %17s %19s %9s\n", "NPCs", "%active", "Polling FSM (s)", "Spikeling LIF (s)", "Speedup");
    for (size_t ci = 0; ci < sizeof(npc_counts)/sizeof(int); ci++) {
        for (size_t fi = 0; fi < sizeof(fractions)/sizeof(double); fi++) {
            int n = npc_counts[ci];
            double f = fractions[fi];
            double t_poll = run_polling(n, f, ticks);
            double t_spike = run_spikeling(n, f, ticks);
            double speedup = t_spike > 0 ? t_poll / t_spike : 0.0;
            printf("%6d %7.0f%% %17.4f %19.4f %8.2fx\n", n, f*100, t_poll, t_spike, speedup);
        }
        printf("\n");
    }
    return 0;
}
